src/status_im/contexts/wallet/effects.cljs

```diff
@@ -17,3 +17,9 @@
       {:MnemonicPhrase phrase
        :paths          paths}
       on-success))))
+
+(rf/reg-fx
+ :effects.common/apply-fn
+ (fn [[f & args]]
+   (when (fn? f)
+     (apply f args))))
```

Comment:
I created this effect so that we can pass callback functions into event handlers, and then those event handlers can return data for describing how to call a function. I think this makes writing unit tests much simpler when passing the callbacks to those event handlers.

But we may not need this effects handler if we decide to always pass on-success or on-error vectors instead. Or we can just call the callbacks directly, but that seems to be more difficult to test properly.

Reply:
:thinking: I'm not sure if this approach is good.

For writing unit test it may be ok since most of the time we want to write concise code, but I'm not sure if this is a desired patter in the app.

In this effect, are we expecting that `f` is a function with side effects?
Could you please give an example of how to use it in our app (besides tests)?

btw, It's really good to explore creative ways of solving problems :+1:

Reply:
Yeah I'm not sure if this a good pattern either, but it allows for us to provide callbacks when dispatching events and it helps with calling those callbacks with the Re-Frame effects system. So it's a general effect for calling functions that likely have some effects inside of them. 

For example, we could create a re-usable event handler and customise the `on-success` and `on-error` (if applicable) logic by passing functions. These functions would be responsible for receiving the results and continuing the chain of events by calling more dispatches.

We can sorta see this pattern being used in this component where functions are passed to continue the chain of events for a re-usable event handler: https://github.com/status-im/status-mobile/blob/9116f5d6a68e38f66158dc79d66cb91180b8b9d6/src/status_im/contexts/settings/wallet/keypairs_and_accounts/import_seed_phrase/view.cljs#L12-L66

However in that example we don't use `:effects.common/apply-fn` because we already have another effect that is responsible for calling those `on-success` and `on-error` functions.

---

All that being said, I'm not strongly in favour of this approach, but I would like a solution for defining re-usable event handlers and providing custom `on-success` or `on-error` logic. I see we already do something like this for the Re-Frame rpc effects, so maybe we grow that pattern so we can share more event handler code.

Definitely want to hear your thoughts on this more too 🙏

Reply:
It'd be interesting if you could show a real word kind of usage, just to understand a bit better how you would use it. 
At first sight, it looks like it loses a bit of the re-frame explicitness, but I am pretty sure I haven't fully grasped how you would use it :). 
With regards of testing as well it'd be good to show an example as well, just to understand better what you mean.

Reply:
> It'd be interesting if you could show a real word kind of usage, just to understand a bit better how you would use it. At first sight, it looks like it loses a bit of the re-frame explicitness, but I am pretty sure I haven't fully grasped how you would use it :). With regards of testing as well it'd be good to show an example as well, just to understand better what you mean.

Yeah I think there's some extra indirection that's being added here which makes it less explicit, but here are some examples that sorta show some of the benefits I think.

Here's an example of some event handlers being defined, in this case it's the `make-seed-phrase-keypair-fully-operable` events. What happens is that an `on-success` and `on-error` callback is passed to the event handler and then an RPC effect is returned. The RPC effect will dispatch `on-success` event and the `on-success` callback will be given to the success event. Then when the success event is called the event handler will decide when to call the on-success callback. This is where `:effects.common/apply-fn` is being used so we avoid calling the `on-success` callback inside the event handler.

https://github.com/status-im/status-mobile/blob/9116f5d6a68e38f66158dc79d66cb91180b8b9d6/src/status_im/contexts/settings/wallet/events.cljs#L92-L120

And here's the tests for these event handlers, which show how an event handler can declare that it wants to call a function as an effect. From a testing perspective, this makes the function call declarative and it allows for us to compare data structures.

https://github.com/status-im/status-mobile/blob/9116f5d6a68e38f66158dc79d66cb91180b8b9d6/src/status_im/contexts/settings/wallet/events_test.cljs#L24-L65

Reply:
@seanstrom If my memory serves me well, such general abstractions are often discussed and implemented in separate PRs (especially re-frame stuff), and then (almost) every CC is  marked as a reviewer to give them a chance to participate. I have my fair share of concerns about something like `:effects.common/apply-fn` and if it wasn't for @ulisesmac to tag me in another unrelated comment in the PR I wouldn't have seen it 🤷🏼

Reply:
> @seanstrom If my memory serves me well, such general abstractions are often discussed and implemented in separate PRs (especially re-frame stuff), and then (almost) every CC is marked as a reviewer to give them a chance to participate. I have my fair share of concerns about something like `:effects.common/apply-fn` and if it wasn't for @ulisesmac to tag me in another unrelated comment in the PR I wouldn't have seen it 🤷🏼

Happy to get more feedback on this too 🙏 

I added this change because I noticed my code was harder to test correctly, and I wanted to get some feedback about whether to write `on-success` and `on-error` handlers as vectors or as functions. 

`:effects.common/apply-fn` is more of a testing utility for the examples I've linked, but we may not need it if we have another way. In general, I have no problem with deferring/removing this change, but I'd still like to get some initial feedback/consensus how to define continuations for events/effects and how to define re-usable event handlers.

Reply:
Just a heads up guys: @seanstrom and I talked about all these things surrounding the `apply-fn` and some more. Sean will try to experiment with a few different ideas and see what comes out of them in this PR.

Thanks for being so open in the discussion 👏🏼